src/sentry/release_health/duplex.py

```diff
@@ -0,0 +1,980 @@
+        else:
+            organization = self._org_from_projects(project_ids)
+
+        return self._dispatch_call(  # type: ignore
+            "get_current_and_previous_crash_free_rates",
+            should_compare,
+            rollup,
+            organization,
+            schema,
+            project_ids,
+            current_start,
+            current_end,
+            previous_start,
+            previous_end,
+            rollup,
+            org_id,
+        )
+
+    def get_release_adoption(
+        self,
+        project_releases: Sequence[ProjectRelease],
+        environments: Optional[Sequence[EnvironmentName]] = None,
+        now: Optional[datetime] = None,
+        org_id: Optional[OrganizationId] = None,
+    ) -> ReleasesAdoption:
+        rollup = self.DEFAULT_ROLLUP  # not used
+        schema = {
+            "adoption": ComparatorType.Ratio,
+            "session_adoption": ComparatorType.Ratio,
+            "users_24h": ComparatorType.Counter,
+            "project_users_24h": ComparatorType.Counter,
+            "project_sessions_24h": ComparatorType.Counter,
+        }
+        should_compare = datetime.utcnow() - timedelta(hours=24) > self.metrics_start
+
+        if org_id is not None:
+            organization = self._org_from_id(org_id)
+        else:
+            organization = self._org_from_projects(project_releases)
+
+        return self._dispatch_call(  # type: ignore
+            "get_release_adoption",
+            should_compare,
+            rollup,
+            organization,
+            schema,
+            project_releases,
+            environments,
+            now,
+            org_id,
+        )
+
+    def run_sessions_query(
+        self,
+        org_id: int,
+        query: QueryDefinition,
+        span_op: str,
+    ) -> SessionsQueryResult:
+        rollup = query.rollup
+
+        schema_for_totals = {
+            "sum(session)": ComparatorType.Counter,
+            "count_unique(user)": ComparatorType.Counter,
+            "avg(session.duration)": ComparatorType.Quantile,
+            "p50(session.duration)": ComparatorType.Quantile,
+            "p75(session.duration)": ComparatorType.Quantile,
+            "p90(session.duration)": ComparatorType.Quantile,
+            "p95(session.duration)": ComparatorType.Quantile,
+            "p99(session.duration)": ComparatorType.Quantile,
+            "max(session.duration)": ComparatorType.Quantile,
+        }
+        schema_for_series = {field: [comparator] for field, comparator in schema_for_totals.items()}
+
+        schema = {
+            "start": ComparatorType.DateTime,
+            "end": ComparatorType.DateTime,
+            "intervals": [ComparatorType.DateTime],
+            "groups": ListSet(
+                schema={
+                    "by": ComparatorType.Ignore,
+                    "series": schema_for_series,
+                    "totals": schema_for_totals,
+                },
+                index_by="by",
```

Comment:
Does this actually work? `groups[0]["by"]` is itself a dict, so you won't be able to use it as a dict key here:
https://github.com/getsentry/sentry/pull/29731/files#diff-90f1dc7390bf824e8bbaf2c0532cf4416195cf9134733a24064b2ec152cc4e87R380

Might need to set `index_by=lambda x: tuple(sorted(x["by"].items()))` like here:

https://github.com/getsentry/sentry/blob/9a0c39c9870cbd426e250dc5ffab7875bb601e72/tests/snuba/api/endpoints/test_organization_sessions.py#L19-L20

Reply:
thanks, didn't relaise that "by" is a dict (thougnt is a string)